Add io.github.sigmasd.VirtAudio

sigmaSd opened the pull request:
<!-- ⚠️⚠️ Submission pull request MUST be made against the `new-pr` **base branch** ⚠️⚠️  -->

<!-- 💡 Go to the preview tab to click the links below 💡 -->

### Please confirm your submission meets all the criteria

<!-- 💡 Please replace each `[ ]` with `[X]` when the step is complete 💡 -->

- [x] Please describe the application briefly. Stream Audio from remoete browsers to your computer as virtual microphones
- [x] The domain used for the application ID is [controlled by the application developer(s)][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read and followed all the [Submission requirements][reqs] and the [Submission guide][reqs2].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor to the project. If not, I contacted upstream developers about this submission. **Link:**

<!-- ⚠️⚠️ DO NOT modify anything below this line ⚠️⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[reqs2]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

hfiguiere commented on io.github.sigmasd.VirtAudio.yml:
```diff
@@ -0,0 +1,267 @@
+finish-args:
+  # Display
+  - --socket=wayland
+  - --socket=x11
```
it's either `x11` or `wayland` + `fallback-x11`

sigmaSd replied:
changed

hfiguiere commented on io.github.sigmasd.VirtAudio.yml:
```diff
@@ -0,0 +1,267 @@
+app-id: io.github.sigmasd.VirtAudio
+runtime: org.freedesktop.Platform
+runtime-version: "23.08"
```
current is 24.08

sigmaSd replied:
changed

hfiguiere commented on io.github.sigmasd.VirtAudio.yml:
```diff
@@ -0,0 +1,267 @@
+  # Network access for local server and WebSocket connections
+  - --share=network
+  # Needed for libinput
+  - --device=input
```
this is no allowed, no one supports it.

sigmaSd replied:
removed

hfiguiere commented:
bot, build io.github.sigmasd.VirtAudio

bbhtt commented on io.github.sigmasd.VirtAudio.yml:
```diff
@@ -0,0 +1,265 @@
+  - --share=network
+  # Needed for mesa
+  - --device=dri
+  # Allow creating temporary files (for creating virtual mics)
+  - --filesystem=/tmp:create
```
there's /tmp inside the sandbox, why do you need host /tmp?

bbhtt replied:
try setting `--env=TMPDIR=/var/tmp`.

sigmaSd replied:
I just noticed this comment, no unfortunately changing to that flag results in the same problem, the audio is not passed to the microphone even though there are no errors when writing to it or errors from ffmpeg or errors in the logs

 I forgot to paste the whole snippet, this is the part that uses the fifo file
```ts
    const result = await new Deno.Command("pactl", {
      args: [
        "load-module",
        "module-pipe-source",
        `source_name=${micName}`,
        `file=${virtualMicPipePath}`,
        "format=s16le",
        "rate=48000",
        "channels=2",
      ],
    }).output();
```

sigmaSd replied:
I'm not sure how to true is this, but asked claude:

>The filesystem permission is needed because PulseAudio needs to access the FIFO file from outside the Flatpak sandbox while your application needs to access it from inside the sandbox. This cross-sandbox communication is what necessitates the filesystem permission.

sigmaSd replied:
I also tried changing the path to a location inside the sandbox but it have the same issue
```ts
    const virtualMicPipePath = `${
      Deno.env.get("XDG_RUNTIME_DIR")
    }/virtual_mic_pipe${this.#id}`;
```

bbhtt replied:
can you check what file it is trying to access? Trying running strace on it. This is proabably influenced by some host puleseaudio configuration that places some socket in `/tmp` only.

sigmaSd replied:
/tmp is not special, any path can work as long as I have permission to it from the sandbox, this also works for example
```ts
    const virtualMicPipePath = `${
      Deno.env.get("XDG_RUNTIME_DIR")
    }/mics/virtual_mic_pipe${this.#id}`;
```
```yml
  - --filesystem=xdg-run/mics:create
```

I think claude have the correct insight, this  cross read/write from inside and outside the sandbox requires this permission

bbhtt replied:
that's better than `/tmp`.

bbhtt replied:
does it need to be :create? You can create the directory in your code and give :ro only?

sigmaSd replied:
Yes :ro works as well, though I don't know why it works, because I do write audio data to the virtual mics inside that directory

sigmaSd replied:
is there a preference if I should choose /tmp or XDG_RUNTIME_DIR, or it doesn't matter I just choose one?

bbhtt replied:
use `xdg-run/mics:ro`

sigmaSd replied:
Turns out :create is needed, it only worked earlier because the folder is already created, I want to change the folder name to virtual_mics, and it turns out it can't create it even tough I do use mkdir  in my app

So I'll just change to xdg-run/virtual_mics:create

sigmaSd replied:
I see now that the app create the directory inside the sandbox only, but the directory need to exist on the host for pulseaudio to work and this is achieved with :create flag

sigmaSd replied:
done, changed to xdg-run/virtual_mics:create

bbhtt commented on io.github.sigmasd.VirtAudio.yml:
```diff
@@ -0,0 +1,265 @@
+          url-template: https://www.freedesktop.org/software/libevdev/libevdev-$version.tar.xz
+    cleanup:
+      - /bin
+      - /include
+      - /lib/pkgconfig
+      - /share/man
```
add these to top level cleanup `*.a` too. no need to repeat.

sigmaSd replied:
done

sigmaSd commented:
bot, build io.github.sigmasd.VirtAudio

sigmaSd commented:
bot, build io.github.sigmasd.VirtAudio

sigmaSd commented:
bot, build io.github.sigmasd.VirtAudio

sigmaSd commented:
The changes I did:
- make deno compile the app into a single executable, this require 2 patches, that I carry here
- vendor registry files, because these files are dynamic they get updated with new versions of packages, which don't care about in this case
- merge common cleanups and make them global

bbhtt commented on registries/slint-ui-binary-darwin-arm64-registry.json:
are these supposed to be here?

sigmaSd replied:
Deno expects these when analyzing npm modules, I tried initially downloading them in flatpak, by specifing the url like this https://registry.npmjs.org/undici-types

This though is not possible because these urls are dynmaic, and they get updated from time to time (with new version of the package) so the sha256 will be different each time

Deno only wants a snapshot at a given time that include the used version

So the only idea I found is fetching and vendoring them here

sigmaSd replied:
i add a commit that decrease their size a lot, and makes them just a couple of lines and human readable

sigmaSd commented:
bot, build io.github.sigmasd.VirtAudio

sigmaSd commented:
bot, build io.github.sigmasd.VirtAudio

sigmaSd commented:
bot, build io.github.sigmasd.VirtAudio

bbhtt commented:
can you try screen recording to show how it works?

sigmaSd commented:
I updated the app to slint nightly, which fixes wayland

@bbhtt I added a demo to the readme https://github.com/sigmaSd/virtAudio?tab=readme-ov-file#demo

sigmaSd commented:
https://github.com/sigmaSd/virtAudio?tab=readme-ov-file#demo

sigmaSd commented:
bot, build io.github.sigmasd.VirtAudio

sigmaSd commented:
bot, build io.github.sigmasd.VirtAudio

bbhtt commented:
/merge
